File: src/82. Remove Duplicates from Sorted List II/main.go
// Solved by @kitanoyoru
// https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func deleteDuplicates(head *ListNode) *ListNode {
  if head == nil {
    return head
  }

  temp := &ListNode{0, head}
  curr, prev := head, temp

  for curr != nil {
    if curr.Next != nil && curr.Val == curr.Next.Val {
      curr = curr.Next
      continue
    }

    if prev.Next != curr {
      prev.Next = curr.Next
    } else {
      prev = prev.Next
    }

    curr = curr.Next
  }

  return temp.Next
}

File: src/82. Remove Duplicates from Sorted List II/main.java
// Solved by @kitanoyoru
// https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
  public ListNode deleteDuplicates(ListNode head) {
    ListNode temp = new ListNode(0, head);
    
    ListNode curr = head, prev = temp;

    while (curr != null) {
      if (curr.next != null && curr.val == curr.next.val) {
        curr = curr.next;
        continue;
      }

      if (prev.next != curr) {
        prev.next = curr.next;
      } else {
        prev = prev.next;
      }

      curr = curr.next;
    }

    return temp.next;
  }
}
